idx/memory/stats.go

```diff
@@ -0,0 +1,11 @@
+package memory
+
+import "github.com/grafana/metrictank/stats"
+
+var (
+	// how many times a corrupt index has been detected
+	CorruptIndex = stats.NewCounter32("idx.memory.corrupt-index")
+
+	// how many times an invalid tag has been found inside the index
+	InvalidTagInIndex = stats.NewCounter32("idx.memory.invalid-tag-in-index")
```

Comment:
we currently log errors for various corrupt index conditions.
```
memory/memory.go
465:				log.Error(3, "memory-idx: node %s missing. Index is corrupt.", n.Path+"."+child)
499:			log.Error(3, "memory-idx: node %s missing. Index is corrupt.", branch)
519:			log.Error(3, "memory-idx: branch %s has no children while trying to delete %s. Index is corrupt", branch, nodes[i])
524:			log.Error(3, "memory-idx: %s not in children list for branch %s. Index is corrupt", nodes[i], branch)

```

my concern is it's already easy enough to ignore these cases, but by using logs for some conditions, and metrics for others, it gets even more easy to neglect this.
so how about we log also instead of introducing new metrics?

Reply:
personally, in all "this should never happen cases. but when it does, it's pretty bad and we should look into it" cases, I think it's fair for MT to crash. as long as we're sure it should never happen due to prior validation etc.
but I know this is not a popular opinion. what do you think? logging errors is the next best thing, but then we leave the burden of tracking the errors up to the operator (e.g. to us)

Reply:
I agree that it would be better to rely either on logs or metrics. Generally I'd prefer to have either just metrics, or metrics and some additional log statements, but no case without metrics. 
That way we can setup alerts depending on conditions, and we can easily get an overview about how all the MTs are behaving in situations where many of them are running. If additional info is required to debug an issue, then that's what a log statement should be used for in addition to the metrics.
I prefer to not have the MT crash. Imagine the user behavior changes in a way that triggers a bug that we were not aware of, do we then want to be alerted while MT tries to perform as good as possible or do we want everything to just stop and the service being down?

Reply:
We should never crash in a case where we can continue in a relatively sane way, bubbling an error up to the caller and/or logging an error if there is a scenario we don't expect is fine, causing an outage because something weird happened is not fine.

Reply:
ok sounds good . so let's then agree to log in all these cases detailing the corrupt data. and if you want to inc a metric, also add that metric to the already existing cases.

Reply:
done: https://github.com/grafana/metrictank/commit/a677649d7631aaf945a66df3f7fb26e4dffe2a98